sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+    /// <param name="loggedHeaderNames"></param>
+    /// <param name="loggedQueryParameters"></param>
+    /// <param name="requestIdHeaderName"></param>
+    /// <param name="logContent"></param>
+    /// <param name="maxLength"></param>
+    /// <param name="assemblyName"></param>
+    public ClientLoggingPolicy(
+        bool isLoggingEnabled = true,
+        List<string>? loggedHeaderNames = default,
+        List<string>? loggedQueryParameters = default,
+        string? requestIdHeaderName = default,
+        bool logContent = false,
+        int maxLength = 4 * 1024,
+        string? assemblyName = default)
+    {
+        LoggedHeaderNames = loggedHeaderNames ?? new List<string>();
+        LoggedQueryParameters = loggedQueryParameters ?? new List<string>();
+        _logContent = logContent;
+        _maxLength = maxLength;
+        _assemblyName = assemblyName;
+        _clientRequestIdHeaderName = requestIdHeaderName;
+        _isLoggingEnabled = isLoggingEnabled;
+    }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public static ClientLoggingPolicy Default { get; } = new ClientLoggingPolicy();
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public List<string> LoggedHeaderNames { get; internal set; }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    public List<string> LoggedQueryParameters { get; internal set; }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    internal PipelineMessageSanitizer? Sanitizer { get; set; }
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            ProcessNext(message, pipeline, currentIndex);
+            return;
+        }
+
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+    }
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            return;
+        }
+
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+    }
+
+    internal virtual void LogRequest()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    internal virtual void LogResponse()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        PipelineRequest request = message.Request;
+        string requestId = string.IsNullOrEmpty(_clientRequestIdHeaderName) ? string.Empty : GetRequestIdFromHeaders(request.Headers, _clientRequestIdHeaderName!);
```

Comment:
Do we want to log an empty string here?

I don't have a good sense of what event shapes look like or whether this would be expected or not, but it seems a little strange to log an empty string into an event parameter to me.  Maybe there is a good reason to do it?  But it might also be nice to detect the case where we're not logging a request id and not add it to the event parameters.

Reply:
The event parameters are ordered to fill in their corresponding part in the event message, for example below the request id parameter goes into the 0 spot. It's a little confusing but the call to WriteEvent doesn't correspond to the attribute. To remove it entirely for certain calls it would have to be defined as a new event.
```csharp
[Event(RequestEvent, Level = EventLevel.Informational, Message = "Request [{0}] {1} {2}\r\n{3}client assembly: {4}")]
    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026", Justification = "WriteEvent is used with primitive types.")]
    public void Request(string requestId, string method, string uri, string headers, string? clientAssembly)
    {
        WriteEvent(RequestEvent, requestId, method, uri, headers, clientAssembly);
    }
```

In message messaging we just log string.empty when we don't have a param but I'll follow up on if that's the general pattern.